src/entity/effect/EffectContainer.php

```diff
@@ -0,0 +1,239 @@
+				$hook($effect);
+			}
+
+			$this->recalculateEffectColor();
+		}
+	}
+
+	/**
+	 * Returns the effect instance active with the specified ID, or null if does not have the
+	 * effect.
+	 */
+	public function get(Effect $effect) : ?EffectInstance{
+		return $this->effects[spl_object_id($effect)] ?? null;
+	}
+
+	/**
+	 * Returns whether the specified effect is active.
+	 */
+	public function has(Effect $effect) : bool{
+		return isset($this->effects[spl_object_id($effect)]);
+	}
+
+	/**
+	 * In the following cases it will return true:
+	 * If a weaker effect of the same type is already applied.
+	 * If a weaker or equal-strength effect is already applied but has a shorter duration.
+	 *
+	 * @return bool whether the effect can been applied.
+	 */
+	public function canAdd(EffectInstance $effect) : bool{
+		$index = spl_object_id($effect->getType());
+		if(isset($this->effects[$index])){
+			$oldEffect = $this->effects[$index];
+			if(
+				abs($effect->getAmplifier()) < $oldEffect->getAmplifier()
+				|| (abs($effect->getAmplifier()) === abs($oldEffect->getAmplifier()) && $effect->getDuration() < $oldEffect->getDuration())
+			){
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Adds an effect.
+	 * If {@link EffectContainer::canAdd()} conditions are met.
+	 *
+	 * @return bool whether the effect has been successfully applied.
+	 */
+	public function add(EffectInstance $effect) : bool{
+		if($this->canAdd($effect)){
+			$index = spl_object_id($effect->getType());
+			$replacesOldEffect = isset($this->effects[$index]);
+
+			$this->effects[$index] = $effect;
+			foreach($this->effectAddHooks as $hook){
+				$hook($effect, $replacesOldEffect);
+			}
+
+			$this->recalculateEffectColor();
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @phpstan-param \Closure(EffectInstance) : bool $effectValidator
+	 */
+	public function setEffectValidatorForBubbles(\Closure $effectValidator) : void{
+		Utils::validateCallableSignature(new CallbackType(new ReturnType(BuiltInTypes::BOOL), new ParameterType("effect", EffectInstance::class)), $effectValidator);
+		$this->effectValidatorForBubbles = $effectValidator;
+	}
+
+	/**
+	 * Recalculates the potion bubbles colour based on the active effects.
+	 */
+	protected function recalculateEffectColor() : void{
+		/** @var Color[] $colors */
+		$colors = [];
+		$ambient = true;
+		foreach($this->effects as $effect){
+			if(($this->effectValidatorForBubbles)($effect)){
+				$level = $effect->getEffectLevel();
+				$color = $effect->getColor();
+				for($i = 0; $i < $level; ++$i){
+					$colors[] = $color;
+				}
+
+				if(!$effect->isAmbient()){
+					$ambient = false;
+				}
+			}
+		}
+
+		if(count($colors) > 0){
+			$this->bubbleColor = Color::mix(...$colors);
+			$this->onlyAmbientEffects = $ambient;
+		}else{
+			$this->bubbleColor = null;
+			$this->onlyAmbientEffects = false;
+		}
+	}
+
+	public function getDefaultBubbleColor() : Color{
+		return $this->defaultBubbleColor;
+	}
```

Comment:
I don't like this. It's only required by AreaEffectCloud.

Really we should check if `getBubbleColor()` is null and then fill it with a default on AreaEffectCloud (or, for the sake of BC, check if the RGBA colour code is 0).

Reply:
I hate having a RGBA(0) and having to do a check on network layer to display the correct color, it would consufuse plugin-devs when doing `getBubbleColor()` on water `AreaEffectCloud`s. but is the only way for BC :c

Reply:
I mean I really prefer the current solution, EffectManager in the first place shouldn't be setting a RGBA(0) when there are no effects

Reply:
> EffectManager/Container in the first place shouldn't be setting a RGBA(0) when there are no effects

Agreed, but I don't think adding more conditionally useless stuff to its API is a good solution. We're gonna have to BC break at some point to fix this anyway.

SplashPotion already suffers from this issue anyway as seen here: https://github.com/pmmp/PocketMine-MP/blob/012894764c1e87569a893014db3ad8e8223f0817/src/entity/projectile/SplashPotion.php#L83-L95

Reply:
To be honest, looking at SplashPotion I'm not really sure why we want to do this EffectContainer refactor anyway. I'm doubtful it makes sense to allow mutating the effects given the increase in complexity it causes.